### PR TITLE
[Newgrounds] Add NewgroundsUser and improve extractor

### PR DESCRIPTION
**Pull request by u-spec-png**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

### CODE REVIEW

File: yt_dlp/extractor/newgrounds.py [hunk 1]
```diff
@@ -88,10 +92,10 @@ def _real_extract(self, url):
         webpage = self._download_webpage(url, media_id)
 
         title = self._html_search_regex(
-            r'<title>([^>]+)</title>', webpage, 'title')
+            r'<title>(.+?)</title>', webpage, 'title')
```
Comment (pukkandan) [inline yt_dlp/extractor/newgrounds.py:95]: why?

File: yt_dlp/extractor/newgrounds.py [hunk 2]
```diff
@@ -128,20 +132,26 @@ def _real_extract(self, url):
...
+            r'"filesize"\s*:\s*["\']?([\d]+)["\']?,', webpage, 'filesize',
             default=None))
+
+        video_type_description = self._html_search_regex(
+            r'"description"\s*:\s*["\']?([^"\']+)["\']?,', webpage, 'filesize',
+            default=None)
+
         if len(formats) == 1:
-            formats[0]['filesize_approx'] = filesize_approx
+            formats[0]['filesize'] = filesize
```
Comment (pukkandan) [inline yt_dlp/extractor/newgrounds.py:152]: are you sure?

File: yt_dlp/extractor/newgrounds.py [hunk 3]
```diff
@@ -128,20 +132,26 @@ def _real_extract(self, url):
...
 
-        if '<dd>Song' in webpage:
+        if 'Audio File' == video_type_description:
```
Comment (pukkandan) [inline yt_dlp/extractor/newgrounds.py:154]:
```suggestion
        if video_type_description == 'Audio File':
```

File: yt_dlp/extractor/newgrounds.py [hunk 4]
```diff
@@ -202,7 +213,57 @@ def _real_extract(self, url):
...
+            'id': 'brian-beaton',
+        },
+        'playlist_mincount': 10,
+    }]
+
+    def _fetch_page(self, channel_id, url, page):
+        page += 1
+        posts_info = self._download_json(
+            f'{url}/page/{page}', channel_id,
+            note=f'Downloading page {page}', errnote=f'Failed to downloading page {page}', headers={
```
Comment (pukkandan) [inline yt_dlp/extractor/newgrounds.py:250]:
```suggestion
            note=f'Downloading page {page}', errnote=f'Failed to download page {page}', headers={
```
though the errnote is unnecessary here imo

File: yt_dlp/extractor/newgrounds.py [hunk 5]
```diff
@@ -202,7 +213,57 @@ def _real_extract(self, url):
...
+                'Accept': 'application/json, text/javascript, */*; q = 0.01',
+                'X-Requested-With': 'XMLHttpRequest',
+            })
+        sequence = posts_info.get('sequence', [])
+        for year in sequence:
+            posts = try_get(posts_info, lambda x: x['years'][f'{year}']['items'])
```
Comment (pukkandan) [inline yt_dlp/extractor/newgrounds.py:256]: why `f'{year}'`? use just `year` or `str(year)`

File: yt_dlp/extractor/newgrounds.py [hunk 6]
```diff
@@ -202,7 +213,57 @@ def _real_extract(self, url):
...
+            for post in posts:
+                path, media_id = re.search(
+                    r'<a[^>]+\bhref=["\'][^"\']+((?:portal/view|audio/listen)/(\d+))[^>]+>',
+                    post).groups()
```
Comment (pukkandan) [inline yt_dlp/extractor/newgrounds.py:260]: `self._search_regex`

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/newgrounds.py [hunk 7]
```diff
@@ -202,7 +213,57 @@ def _real_extract(self, url):
...
                 self.url_result(
-                    'https://www.newgrounds.com/%s' % path,
+                    f'https://www.newgrounds.com/{path}',
                     ie=NewgroundsIE.ie_key(), video_id=media_id))
 
         return self.playlist_result(entries, playlist_id, title)
+
+
+class NewgroundsUserIE(InfoExtractor):
+    _VALID_URL = r'https?://(?P<id>[^\.]+)\.newgrounds\.com/(movies|audio)'
```
Comment (pukkandan) [inline yt_dlp/extractor/newgrounds.py:223]:
```suggestion
    _VALID_URL = r'https?://(?P<id>[^\.]+)\.newgrounds\.com/(?:movies|audio)/?(?:[#?]|$)'
```
should fix the tests

File: yt_dlp/extractor/newgrounds.py (hunk 2 above)
Comment (u-spec-png) [reply yt_dlp/extractor/newgrounds.py:152]: Yes, here are files:
```
$ du -b ~/Downloads/*.mp3
4557743	/home/pogaca/Downloads/B7 - BusMode - 549479.mp3
2192718	/home/pogaca/Downloads/Zombie Wind - 500392.mp3
```
File sizes detected by yt-dlp:
file1
file2

File: yt_dlp/extractor/newgrounds.py (hunk 1 above)
Comment (u-spec-png) [reply yt_dlp/extractor/newgrounds.py:94]: Title could contain `>` character and not be detected by yt-dlp.

Comment (nyuszika7h) [reply yt_dlp/extractor/newgrounds.py:94]: That's not valid HTML. It would be escaped as `&gt;`.

Comment (u-spec-png) [reply yt_dlp/extractor/newgrounds.py:94]: I didn't know that, still gonna leave it doesn't break anything.

Comment (pukkandan) [reply yt_dlp/extractor/newgrounds.py:94]: if the webpage contains multiple title tags, it could

File: yt_dlp/extractor/newgrounds.py [hunk 8]
```diff
@@ -197,7 +196,7 @@ def _real_extract(self, url):
         webpage = self._download_webpage(url, playlist_id)
 
         title = self._search_regex(
-            r'<title>([^>]+)</title>', webpage, 'title', default=None)
+            r'<title>(.+?)</title>', webpage, 'title', default=None)
```
Comment (pukkandan) [inline yt_dlp/extractor/newgrounds.py:199]:
```suggestion
            r'<title>([^>]+)</title>', webpage, 'title', default=None)
```

**Pull request merged**